escrow/src/main.sw

```diff
@@ -1,257 +1,346 @@
+    #[storage(read, write)]fn propose_arbiter(arbiter: Arbiter, identifier: u64) {
+        // The assertions ensure that only the seller can propose a new arbiter and the arbiter
+        // cannot be the buyer / seller, the arbiter will be able to take a none-zero payment
 
-                storage.seller.deposited = false;
-                storage.seller.approved = false;
+        let escrow = storage.escrows.get(identifier);
 
-                transfer_to_output(storage.asset_amount, storage.asset, storage.seller.address);
-            }
-        } else {
-            revert(0);
-        };
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+
+        let user = msg_sender().unwrap();
+
+        require(user == escrow.seller.address, UserError::Unauthorized);
+        require(arbiter.address != escrow.buyer.address, ArbiterInputError::CannotBeBuyer);
+        require(arbiter.address != escrow.seller.address, ArbiterInputError::CannotBeSeller);
+        require(0 < arbiter.fee_amount, ArbiterInputError::FeeCannotBeZero);
+        require(arbiter.fee_amount == msg_amount(), ArbiterInputError::FeeDoesNotMatchAmountSent);
+        require(arbiter.asset == msg_asset_id(), ArbiterInputError::AssetDoesNotMatch);
+
+        // If there is a previous proposal then we must transfer those funds back to the seller
+        let proposal = storage.arbiter_proposal.get(identifier);
+        if proposal.is_some() && 0 < proposal.unwrap().fee_amount {
+            transfer(proposal.unwrap().fee_amount, proposal.unwrap().asset, escrow.seller.address);
+        }
 
-        true
+        storage.arbiter_proposal.insert(identifier, Option::Some(arbiter));
+
+        log(ProposedArbiterEvent {
+            arbiter, identifier
+        });
+    }
+
+    #[storage(read, write)]fn resolve_dispute(identifier: u64, payment_amount: u64, user: Identity) {
+        // The assertions ensure that a resolution can only occur during a dispute and only once
+        // by the specified arbiter. The deposit will be sent to either the buyer or seller and the
+        // arbiter can choose their payment amount up to the deposit from the seller
+
+        let mut escrow = storage.escrows.get(identifier);
+
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+        require(escrow.disputed, StateError::NotDisputed);
+        require(msg_sender().unwrap() == escrow.arbiter.address, UserError::Unauthorized);
+        require(user == escrow.buyer.address || user == escrow.seller.address, UserInputError::InvalidRecipient);
+        require(escrow.buyer.asset.is_some() && 0 < escrow.buyer.deposited_amount, StateError::CannotResolveBeforeDesposit);
+        require(payment_amount <= escrow.arbiter.fee_amount, ArbiterInputError::PaymentTooLarge);
+
+        escrow.state = State::Completed;
+        storage.escrows.insert(identifier, escrow);
+
+        transfer(escrow.buyer.deposited_amount, escrow.buyer.asset.unwrap(), user);
```

Comment:
How does the seller get their original assets back? Who do they go to? Since the `escrow.state` is set to `State::Completed` isn't it locked in the contract forever?

Reply:
When a resolution is performed by the arbiter the arbiter gets to select who the deposit from the buyer goes to i.e. the buyer or seller. Since the arbiter has performed work they get to be rewarded with the collateral from the seller. The arbiter gets to select up to the maximum amount of collateral from the seller and the difference will be sent back to the seller so for example if the seller has some asset of quantity 100 and the arbiter thinks that their work was only worth 80 then the remaining 20 go back to the user (as seen on line 238).

This leaves a potential deposit from the seller for a change of arbiter and thus on line 243 the code checks to see if the seller has made proposal for a change of arbiter (which requires collateral) and if the buyer has not accepted then it's stuck there so we must now send that back.